[RAM] Refactor rules page to use EuiPageTemplate

JiaweiWu opened the pull request:
## Summary
Resolves: https://github.com/elastic/kibana/issues/142374

Refactors the rules page in Management to follow the new `EuiPageTemplate` and `EuiPageTemplate.EmptyPrompt`. The no rules and no permission prompts are centered correctly according to EUI guidelines.

Tested:
- Rules page
- Logs page
- O11Y Rules page

The `create rule` buttons have also been moved to the top right of the header

and all work as intended. However, the O11y rules page empty prompt is not automatically centered since they are structuring their page templates differently.

![emptypromptwithcreate](https://user-images.githubusercontent.com/74562234/206249274-93a4f866-97e9-471b-85c3-c368cdb74a71.png)


![euipagetemplatewithrule](https://user-images.githubusercontent.com/74562234/203172564-48c86855-13ab-4899-9225-7fb4694c8cf3.png)

JiaweiWu commented:
@elasticmachine merge upstream

mdefazio commented:
Is part of this PR to also change the page template when there are Rules? So the create button would be on the right side in the Page header portion?

JiaweiWu commented:
@mdefazio that is correct, I am going to fix that

JiaweiWu commented:
@mdefazio I moved the create rule button, just like how O11Y currently has theirs. Although I am unsure if we want to have the plus icon next to the create button.

XavierM commented:
@JiaweiWu I think we should match the empty prompt like the connector
https://github.com/elastic/kibana/pull/142485#issuecomment-1270467387

XavierM commented on x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rules_list.tsx:
```diff
@@ -995,24 +1005,39 @@ export const RulesList = ({
-      );
-    }
 
+  const showPrompt = noData && !rulesState.isLoading && !ruleTypesState.isLoading;
```
can we create a component outside of this file?

JiaweiWu replied:
done

JiaweiWu commented:
So we want the create button in the empty prompt, but when we have rules we want it in the top right of the header?

JiaweiWu commented on x-pack/plugins/triggers_actions_ui/public/application/home.tsx:
```diff
@@ -78,14 +89,16 @@ export const TriggersActionsUIHome: React.FunctionComponent<RouteComponentProps<
           </span>
         }
         rightSideItems={[
+          ...(createRuleButton ? [createRuleButton] : []),
```
We are actually checking the auth.

So we need to check rule type auth to determine if the user can create rules. I added a new prop `setCreateRuleButton` to the `rulesList` so we can let the `rulesList` determine whether or not we should show this button.

This way we don't have to refactor the entire component to move the rule types fetch to the parent component

JiaweiWu commented:
@elasticmachine merge upstream

JiaweiWu commented:
@XavierM I addressed your comments, the only odd thing here is the documentation link in the header:

![neweuitemplate](https://user-images.githubusercontent.com/74562234/203667832-01a5659e-1624-4580-9e4d-7b5686c9451d.png)

Because the rules page has 2 tabs, if we want to hide the documentation link ONLY for the rules list while the empty prompt is visible, I can change it to do that but the logic gets a bit complicated.

mdefazio reviewed (approved):
LGTM!

guskovaue commented:
Hi, @JiaweiWu !
<img width="1313" alt="Screenshot 2022-12-07 at 09 22 29" src="https://user-images.githubusercontent.com/26089545/206179968-e4d22c4e-f336-44fe-a47c-152286b8552e.png">
<img width="1321" alt="Screenshot 2022-12-07 at 09 22 52" src="https://user-images.githubusercontent.com/26089545/206179983-e493ed69-e642-4ae8-9356-59eb7b9262f1.png">

My concerns are: for o11y, we have now 2 `Create rule` buttons and two `Documentation`. Should we stick with one? 
And also `Create your first rule` is not in the center.

I see in previous comments you mentioned that `the O11y rules page empty prompt is not automatically centered since they are structuring their page templates differently`. But I did not get if we can centered it manually?

And also see you wrote about `only odd thing here is the documentation link in the header`, but do not see any decision made. Maybe not here, but in follow up ticket.

JiaweiWu commented:
> My concerns are: for o11y, we have now 2 `Create rule` buttons and two `Documentation`. Should we stick with one? And also `Create your first rule` is not in the center.

Fixed with my last commit: https://github.com/elastic/kibana/pull/144070/commits/36d9fa63a60798cf36b402c6c5ce7d4196719857

> I see in previous comments you mentioned that `the O11y rules page empty prompt is not automatically centered since they are structuring their page templates differently`. But I did not get if we can centered it manually?

So with the new EUI page template, centering is done automatically, but it depends on the parent component  to implement it correctly: https://elastic.github.io/eui/#/templates/page-template/examples#empty-pages-or-content. So I think it will be up to them to change their plugin to fit the new EUI templates. 

The changes here do not affect existing behaviors, but it supports the new templates. When they decide to make the change, it will be centered. 

> And also see you wrote about `only odd thing here is the documentation link in the header`, but do not see any decision made. Maybe not here, but in follow up ticket.

Good idea, I'll create another ticket :slightly_smiling_face:

JiaweiWu commented:
@elasticmachine merge upstream

guskovaue commented on x-pack/plugins/triggers_actions_ui/public/application/sections/rule_details/components/rule_event_log_list_table.tsx:
```diff
@@ -201,6 +204,12 @@ export const RuleEventLogListTable = <T extends RuleEventLogListOptions>(
 
+  useEffect(() => {
+    setHeaderActions?.([<RulesListDocLink />]);
+    return () => setHeaderActions?.();
```
Why do we need cleanup function?

guskovaue commented:
Hi, @JiaweiWu !
I see you decided not to change rules prompt in o11y. I think it ok for this PR. But maybe in the future we can follow up and make it align between o11y and Management. Maybe follow up ticket? What do you think?

Also I made couple comments. I have doubts we need cleanup functions in useEffect. But maybe I do not know something.

But in general LGTM!

JiaweiWu commented on x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rules_list.tsx:
```diff
@@ -1007,24 +1019,29 @@ export const RulesList = ({
+  }, [initialLoad, showPrompt, authorizedToCreateAnyRules]);
+
+  useEffect(() => {
+    return () => setHeaderActions?.();
```
Mostly for defensive reasons, in case we decide to add another tab, we wouldn't want the old headers to appear in the new tab. Especially because the rules list has a "create rule" button, which is only relevant for the rules list
